### PR TITLE
[GEOS-11209] Proof Key of Code Exchange (PKCE)

### PR DESCRIPTION
**Pull request by jodygarnett**
[GEOS-11209](https://osgeo-org.atlassian.net/browse/GEOS-11209) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=geoserver&utm_campaign=badge_info)

* Add PKCE configuration option for open id connect module
* Include OIDC_LOGGING profile to aid in troubleshooting
* PKCEAuthenticationEntryPoint to generate and store code_verifier, and then include code_challenge in authorization request
* PKCERequestEnhancer to retrieve code_verifier and include in access token request

[GEOS-11212](https://osgeo-org.atlassian.net/browse/GEOS-11212):

* Use JWK URI to check access token

Some protected methods introduced into oauth2-core classes to allow override for PKCE workflow.
  

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [ ] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [x] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [ ] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [x] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [x] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [ ] Bug fixes and small new features are presented as a single commit.
- [x] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

File: doc/en/user/source/community/oauth2/oidc.rst [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+Reference:
+
+* `rfc7636 Proof Key for Code Exchange by OAuth Public Clients <https://datatracker.ietf.org/doc/html/rfc7636>`_
+
+JSON Web Key set URI
+^^^^^^^^^^^^^^^^^^^^
+
+The ``JSON Web Key set URI`` provides the location of a document of public keys that can be used to check the signature of the provided accessToken.
+
+Optional: When using a role service to obtain the roles associated with access token, it is no longer required to use ``Check Token Endpoint URL``. If you leave that field blank you may rely only on the ``JSON Web Key set URI`` signature check.
```
Comment (jodygarnett) [inline doc/en/user/source/community/oauth2/oidc.rst:107]: These are the new docs for PKCE work and JWKS URI use.

**Comment by afabiani** [comment]
@jodygarnett thanks for this update. Overall it looks very good to me. Just two minor comments:

 - A couple of test cases are missing the headers I guess
 - Did you test this code with a custom KID? From what I can see it should work, just to confirm
 - It would be nice (if you have the chance, otherwise I can eventually contribute) to add some more details on the docs about validating an OpenID Connect ID Token, since most of the modern providers requires it nowdays

**Comment by jodygarnett** [comment]
> * It would be nice (if you have the chance, otherwise I can eventually contribute) to add some more details on the docs about validating an OpenID Connect ID Token, since most of the modern providers requires it nowdays

My experience with security topics is limited so I do not feel comfortable offering a clear explanation. 

This pull request only adds validation in one location - where JWKs is being used to check the token. 

The other place GeoServer uses the JWKS URI has a comment indicated it will be used by JwkTokenStore to "verify" ... but I see no evidence of validation being performed by this JwkTokenStore:
```
    private void validate(OAuth2AccessToken token) {
        Object maybeIdToken = token.getAdditionalInformation().get("id_token");
        if (maybeIdToken instanceof String) {
            String idToken = (String) maybeIdToken;
            setAsRequestAttribute(OpenIdConnectAuthenticationFilter.ID_TOKEN_VALUE, idToken);
            // among other things, this verifies the token
            if (store != null) store.readAuthentication(idToken);
            // TODO: the authentication just read could contain roles, could be treated as
            // another role source... but needs to be made available to role computation
        }
    }
```
 
If validation needs to be added here we can make a new issue and pull request (and document how validation works at that time).

**Comment by afabiani** [comment]
> > * It would be nice (if you have the chance, otherwise I can eventually contribute) to add some more details on the docs about validating an OpenID Connect ID Token, since most of the modern providers requires it nowdays
> 
> My experience with security topics is limited so I do not feel comfortable offering a clear explanation.
> 
> This pull request only adds validation in one location - where JWKs is being used to check the token.
> 
> The other place GeoServer uses the JWKS URI has a comment indicated it will be used by JwkTokenStore to "verify" ... but I see no evidence of validation being performed by this JwkTokenStore:
> 
> ```
>     private void validate(OAuth2AccessToken token) {
>         Object maybeIdToken = token.getAdditionalInformation().get("id_token");
>         if (maybeIdToken instanceof String) {
>             String idToken = (String) maybeIdToken;
>             setAsRequestAttribute(OpenIdConnectAuthenticationFilter.ID_TOKEN_VALUE, idToken);
>             // among other things, this verifies the token
>             if (store != null) store.readAuthentication(idToken);
>             // TODO: the authentication just read could contain roles, could be treated as
>             // another role source... but needs to be made available to role computation
>         }
>     }
> ```
> 
> If validation needs to be added here we can make a new issue and pull request (and document how validation works at that time).

Ok yes, I see. Unfortunately yes, letting the store validating the `id_token` most probably will fail in case of a custom `KIDs` I'm afraid.
However I agree that this is not something that must/should be addressed by this PR.
I'll try to open a new issue for that, since this will become a problem for sure in the future.

**Comment by jodygarnett** [comment]
> * Did you test this code with a custom KID? From what I can see it should work, just to confirm

@afabiani I am not sure what this means; can you explain? kid is key identifier, is the an option to specify a different key identifier to look up in JWKs ?!

**Comment by jodygarnett** [comment]
Feedback addressed, rebased and cleaned up commit history. Thanks for feedback @davidblasby and @afabiani .

**Pull request merged**